src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,613 @@
+    public <V> String toJSONString(V value, ValueAccessor<V> accessor, ProtocolVersion protocolVersion)
+    {
+        StringBuilder sb = new StringBuilder();
+        sb.append('[');
+        List<V> split = split(value, accessor);
+        for (int i = 0; i < dimension; i++)
+        {
+            if (i > 0)
+                sb.append(", ");
+            sb.append(elementType.toJSONString(split.get(i), accessor, protocolVersion));
+        }
+        sb.append(']');
+        return sb.toString();
+    }
+
+    @Override
+    public Term fromJSONObject(Object parsed) throws MarshalException
+    {
+        if (parsed instanceof String)
+            parsed = JsonUtils.decodeJson((String) parsed);
+
+        if (!(parsed instanceof List))
+            throw new MarshalException(String.format(
+            "Expected a list, but got a %s: %s", parsed.getClass().getSimpleName(), parsed));
+
+        List<?> list = (List<?>) parsed;
+        if (list.size() != dimension)
+            throw new MarshalException(String.format("List had incorrect size: expected %d but given %d; %s", dimension, list.size(), list));
+        List<Term> terms = new ArrayList<>(list.size());
+        for (Object element : list)
+        {
+            if (element == null)
+                throw new MarshalException("Invalid null element in list");
+            terms.add(elementType.fromJSONObject(element));
+        }
+
+        return new Vectors.DelayedValue<>(this, terms);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        VectorType<?> that = (VectorType<?>) o;
+        return dimension == that.dimension && Objects.equals(elementType, that.elementType);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(elementType, dimension);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toString(false);
+    }
+
+    @Override
+    public String toString(boolean ignoreFreezing)
+    {
+        return getClass().getName() + TypeParser.stringifyVectorParameters(elementType, ignoreFreezing, dimension);
+    }
+
+    private void check(List<?> values)
+    {
+        if (values.size() != dimension)
+            throw new MarshalException(String.format("Required %d elements, but saw %d", dimension, values.size()));
+
+        // This code base always works with a list that is RandomAccess, so can use .get to avoid allocation
+        for (int i = 0; i < dimension; i++)
+        {
+            Object value = values.get(i);
+            if (value == null || (value instanceof ByteBuffer && elementSerializer.isNull((ByteBuffer) value)))
+                throw new MarshalException(String.format("Element at index %d is null (expected type %s); given %s", i, elementType.asCQL3Type(), values));
+        }
+    }
+
+    public abstract class VectorSerializer extends TypeSerializer<List<T>>
```

Comment:
I can do this, but felt it was best to keep in one place rather than move around.  I also found that some types didn't know that we are required to have a 1:1 mapping of `type -> serializer` and `serializer -> type`... which led to fun bug discovery in fuzz testing that could actually corrupt data!

By using an instance level class, I am forced to keep this 1:1 mapping.... I can def maintain with different files, but figured its easier to also break as well.

Reply:
Makes sense to me.